src/metabase/task/send_pulses.clj

```diff
@@ -1,136 +1,197 @@
-
-;;; ------------------------------------------------------ Task ------------------------------------------------------
-
-(defn- monthday [dt]
-  (cond
-    (timepr/first-day-of-month? dt) :first
-    (timepr/last-day-of-month? dt)  :last
-    (= 15 (time/day dt))            :mid
-    :else                           :other))
-
-(defn- monthweek [dt]
-  (let [curr-day-of-month  (time/day dt)
-        last-of-month      (time/day (time/last-day-of-the-month dt))
-        start-of-last-week (- last-of-month 7)]
-    (cond
-      (> 8 curr-day-of-month)                  :first
-      (< start-of-last-week curr-day-of-month) :last
-      :else                                    :other)))
+    (log/infof "Deleting %d PulseChannels with id: %s due to having no recipients" (count ids-to-delete) (str/join ", " ids-to-delete))
+    (t2/delete! :model/PulseChannel :id [:in ids-to-delete])))
 
-(jobs/defjob ^{:doc "Triggers the sending of all pulses which are scheduled to run in the current hour"} SendPulses [_]
-  (try
-    (task-history/with-task-history {:task "send-pulses"}
-      ;; determine what time it is right now (hour-of-day & day-of-week) in reporting timezone
-      (let [reporting-timezone (driver/report-timezone)
-            now                (if (empty? reporting-timezone)
-                                 (time/now)
-                                 (time/to-time-zone (time/now) (time/time-zone-for-id reporting-timezone)))
-            curr-hour          (time/hour now)
-            ;; clj-time produces values of 1-7 here (Mon -> Sun), so we subtract 1 from it to make the values
-            ;; zero-based to correspond to the indices in pulse-channel/days-of-week
-            curr-weekday       (->> (dec (time/day-of-week now))
-                                    (get pulse-channel/days-of-week)
-                                    :id)
-            curr-monthday      (monthday now)
-            curr-monthweek     (monthweek now)]
-        (send-pulses! curr-hour curr-weekday curr-monthday curr-monthweek))
-      (clear-pulse-channels!))
-    (catch Throwable e
-      (log/error e "SendPulses task failed"))))
+(defn- reprioritize-send-pulses!
+  "Find all active pulse Channels, group them by pulse-id and schedule time and create a trigger for each.
 
-(def ^:private send-pulses-job-key     "metabase.task.send-pulses.job")
-(def ^:private send-pulses-trigger-key "metabase.task.send-pulses.trigger")
+  Also remove PulseChannels that don't have any recipients."
+  []
+  (log/info "Reprioritizing send pulses")
+  (clear-pulse-channels!)
+  (let [trigger-slot->pc-ids (as-> (t2/select :model/PulseChannel :enabled true) results
+                               (group-by #(select-keys % [:pulse_id :schedule_type :schedule_day :schedule_hour :schedule_frame]) results)
+                               (update-vals results #(map :id %)))]
+    ;; TODO: use info from task-history to set priority of each pulses of the same time slot
```

Comment:
Is the idea to prioritize pulses by average execution time? It seems a bit like over-engineering while we have opportunities to optimize the pulses themselves by so much

Reply:
It's a nice idea, I just doubt it'll ever be necessary to warrant spending time on it